### ISSUE TITLE
Reading in a large detector from json is slow

### ISSUE THREAD

**Issue by sjoerd-bouma** [nu-radio/NuRadioMC#592]
Reading in a large detector description from a `.json` file is slower than expected; for the LOFAR detector description (O(1000) channels), reading in the json description using
```
det = Detector('LOFAR.json', antenna_by_depth=False)
```
takes O(1 minute). If instead we read in the json first, and then load the detector from the dictionary, this only takes O(1 s):
```
with open('LOFAR.json', 'r') as f:
    det_dict = json.load(f)
det = Detector(json_filename=None, source='dictionary', dictionary=det_dict, antenna_by_depth=False)
```

I'm not sure how the first case is currently implemented - most time is spent in `tinydb_serialization._decode_deep` and `json.decoder.raw_decode`. However, seeing as the second case is (?) equivalent to the first, we should probably change the implementation to just internally read in the json as a dictionary first, and load the detector from that.

**Comment by MijnheerD** [comment]
Something I do not entirely get though, is that on lines 95 and 96 of `generic_detector.py` it seems like the JSON file decoded and the `Detector` is constructed using the dictionary on line 97-99. So would the time be spent on the `Query()` calls then?

**Comment by anelles** [comment]
If I am not fully mistaken @cg-laser was the one to introduce tiny_db in here. Maybe it really is too tiny for what we are doing here? 
I would be open to changing it, if we can make it work seamlessly with what @jakobhenrichs and @fschlueter have been implementing for the DB-based detector.

**Comment by fschlueter** [comment]
@anelles, I do not think so. The RNO-G detector class uses a custom buffering "strategy", I do not think that this is transferable to this detector.

**Comment by cg-laser** [comment]
TinyDB is still used when giving a json object to the detector class. I second your suggestion Sjoerd. It seems to efficiently solve the problem.

**Comment by MijnheerD** [comment]
As a development on this issue, when loading a complete detector description the initialisation is much faster. So I think this is particular issue when trying to load a GenericDetector with lots of references, which is where I would guess it spends most of its time.

**Comment by fschlueter** [comment]
Can this be closed? I think this is a feature of the generic detector class...

**Comment by MijnheerD** [comment]
Probably it can. Just this morning I did come across a package that could help us though: https://pypi.org/project/pysimdjson/ It is a library which supposedly speeds up the parsing of JSON files a lot, thanks to some parallel execution. It does only support Python 3.9+ however, and I think the performance increase depends on the use case so we should test if this also benefits ours. But that could be something for the future :-)

**Comment by sjoerd-bouma** [comment]
The problem is that the generic detector currently uses a filesystem-based cache if the source is json, which is much slower than the memory-based one if a dictionary is used as a source. 

Mostly I think this is just a problem with TinyDB, and my preference is still to replace the GenericDetector implementation with a Python dictionary implementation that gets rid of TinyDB altogether - I believe @PhilippLaub has done quite a bit of work in this direction but I keep forgetting to ask him what the status is.

**Comment by fschlueter** [comment]
oh if we could get rid of tinyDB that would be great but is it easy to do so?

**Comment by PhilippLaub** [comment]
In  the branch `lofar/no-tinydb` I modified the base detector to use dicts instead of TinyDB. If I remember correctly, it was faster for larger detectors.

**Comment by PhilippLaub** [comment]
I think there is a check for commission time.

**Comment by fschlueter** [comment]
@PhilippLaub are you still working on this?

**Comment by PhilippLaub** [comment]
No, I have not worked on this for a year now, but I think it did work for LOFAR. I am not sure if the detector can be saved in an event. What I never tested was the buffering of SQL detectors.
